### PR TITLE
Improve GG PopupMenu Idiom

### PR DESCRIPTION
**Pull request by LGM-Doyle**
***Issue***
The idiom for using `GG::PopupMenu` is:
- add `GG::MenuItem` structures with a specific 'significant' integer index to another `GG::MenuItem`
- then run the `MenuItem` in the `GG::PopupMenu` and match the returned integer in a switch statement to the 'significant' integers used for each `MenuItem` to select handler code for the result.

This leads to several sub-optimal results:
- the setup and handling code for a given menu item are poorly localized and difficult to clearly associate with each other.
- the only data that can be associated with a menu item is the integer.  Multiple menu items in the same context menu have to use a single coding scheme to share the integer.  See the `ObjectsWnd` code for examples difficult coding problems.
- it is difficult to share handlers, because of the coupling introduced by the 'significant' integer.

***Solution***
In this PR I change `MenuItem` to use a `std::function<void()>` as the callback action instead the integer.  There is no switch statement.  The `popup.Run()` statement calls the handler of the selected menu item on exit.

This solves the problems described above:
- the menuitem and handlers can be grouped in any intuitive way because they are not coupled by the switch statement.
- any data can be associated with a meu item handler because it can be passed a named function with `std::bind` and capture arbitrary state with a lambda function.  They are not collectively limited to using a single common integer.
- it is easier to share handlers because there is less artificial coupling. See the production and research queue code for an example of reusing the base class handlers in the derived classes.

### CODE REVIEW

File: UI/BuildingsPanel.cpp [hunk 1]
```diff
@@ -418,56 +418,38 @@ void BuildingIndicator::RClick(const GG::Pt& pt, GG::Flags<GG::ModKey> mod_keys)
...
-                    OrderPtr(new ScrapOrder(empire_id, m_building_id)));
-            break;
-        }
-
-        case 4: { // un-scrap building
-            if (!m_order_issuing_enabled)
-                break;
 
+    auto scrap_building_action = [this, empire_id]() {
+                HumanClientApp::GetApp()->Orders().IssueOrder(OrderPtr(new ScrapOrder(empire_id, m_building_id)));};
```
Comment (geoffthemedio) [inline UI/BuildingsPanel.cpp]: weird indenting and linebreaking...

File: UI/BuildingsPanel.cpp [hunk 2]
```diff
@@ -418,56 +418,38 @@ void BuildingIndicator::RClick(const GG::Pt& pt, GG::Flags<GG::ModKey> mod_keys)
...
+        } else {
+            // create popup menu with "Cancel Scrap" option
+            popup.AddMenuItem(GG::MenuItem(UserString("ORDER_CANCEL_BUIDLING_SCRAP"), false, false, un_scrap_building_action));
         }
     }
+
+    const std::string& building_type = building->BuildingTypeName();
+    const BuildingType* bt = GetBuildingType(building_type);
+    if (bt) {
+        auto pedia_lookup_building_type_action = [building_type]() { ClientUI::GetClientUI()->ZoomToBuildingType(building_type); };
```
Comment (geoffthemedio) [inline UI/BuildingsPanel.cpp]: I'd move the first code line to a new indented line... On the same line as the { looks weird and is hard to read due to it being so far to the right

**Comment by geoffthemedio** [comment]
In general, looks cleaner / easier to use.

**Comment by adrianbroher** [comment]
I agree with @geoffthemedio here. Well done.

But I'm wondering why you named all the lambda handlers.  Given the fact that most of them wrap a single to a few statement you could also moved them into the corresponding `PopupMenu::AddMenuItem` call, completely eliminating the need for a local variable **and** placing the _action_ within the same block of code as the call to give the code a maximum of logic locality.

Also the PR presentation is excellent.

**Comment by geoffthemedio** [comment]
I somewhat like having the local variable names for the handlers... It's easier to read / understand a concise variable name than a block of code, particularly when you get to 8 or so non-trivial lines in the relevant function. They are still defined right next to where they are used, so it's not like you have to search through the code to find the relevant called function. And having a multi-line function definition within a multi-parameter function call is awkward at best.

File: UI/SidePanel.cpp [hunk 3]
```diff
@@ -2041,85 +2034,70 @@ void SidePanel::PlanetPanel::RClick(const GG::Pt& pt, GG::Flags<GG::ModKey> mod_
...
         && m_planet_name->InClient(pt))
     {
-        menu_contents.next_level.push_back(GG::MenuItem(UserString("SP_RENAME_PLANET"), 1, false, false));
+        auto rename_action = [this, planet]() { // rename planet
+            std::string plt_name = planet->Name();
+            CUIEditWnd edit_wnd(GG::X(350), UserString("SP_ENTER_NEW_PLANET_NAME"), plt_name);
+            edit_wnd.Run();
+            if (edit_wnd.Result() != "" && edit_wnd.Result() != planet->Name() && m_order_issuing_enabled)
+            {
+                HumanClientApp::GetApp()->Orders().IssueOrder(OrderPtr(new RenameOrder(HumanClientApp::GetApp()->EmpireID(), planet->ID(), edit_wnd.Result())));
```
Comment (geoffthemedio) [inline UI/SidePanel.cpp]: I'd break this call into two lines... it's awkwardly long

File: UI/SidePanel.cpp [hunk 4]
```diff
@@ -2041,85 +2034,70 @@ void SidePanel::PlanetPanel::RClick(const GG::Pt& pt, GG::Flags<GG::ModKey> mod_
...
+
     if (planet->OwnedBy(HumanClientApp::GetApp()->EmpireID()) && m_order_issuing_enabled
         && m_planet_name->InClient(pt))
     {
-        menu_contents.next_level.push_back(GG::MenuItem(UserString("SP_RENAME_PLANET"), 1, false, false));
+        auto rename_action = [this, planet]() { // rename planet
+            std::string plt_name = planet->Name();
+            CUIEditWnd edit_wnd(GG::X(350), UserString("SP_ENTER_NEW_PLANET_NAME"), plt_name);
+            edit_wnd.Run();
+            if (edit_wnd.Result() != "" && edit_wnd.Result() != planet->Name() && m_order_issuing_enabled)
```
Comment (geoffthemedio) [inline UI/SidePanel.cpp]: possibly also break this over 2+ lines, especially since the { is on a separate line anyway

File: UI/SidePanel.cpp [hunk 5]
```diff
@@ -803,33 +803,26 @@ class SidePanel::SystemNameDropDownList : public CUIDropDownList {
...
-                break;
-            default:
-                break;
+        auto rename_action = [this, system]() {
+            const std::string& old_name(system->Name());
+            CUIEditWnd edit_wnd(GG::X(350), UserString("SP_ENTER_NEW_SYSTEM_NAME"), old_name);
+            edit_wnd.Run();
+            const std::string& new_name(edit_wnd.Result());
+            if (m_order_issuing_enabled && !new_name.empty() && new_name != old_name) {
+                HumanClientApp::GetApp()->Orders().IssueOrder(
```
Comment (geoffthemedio) [inline UI/SidePanel.cpp:815]: this line break is good, IMO

File: UI/SidePanel.cpp [hunk 6]
```diff
@@ -2041,85 +2034,70 @@ void SidePanel::PlanetPanel::RClick(const GG::Pt& pt, GG::Flags<GG::ModKey> mod_
...
 
-    GG::MenuItem menu_contents;
+    CUIPopupMenu popup(pt.x, pt.y);
+
     if (planet->OwnedBy(HumanClientApp::GetApp()->EmpireID()) && m_order_issuing_enabled
         && m_planet_name->InClient(pt))
     {
-        menu_contents.next_level.push_back(GG::MenuItem(UserString("SP_RENAME_PLANET"), 1, false, false));
+        auto rename_action = [this, planet]() { // rename planet
+            std::string plt_name = planet->Name();
```
Comment (geoffthemedio) [inline UI/SidePanel.cpp]: could be a `const std::string&` I think

File: UI/SidePanel.cpp (hunk 4 above)
Comment (geoffthemedio) [reply UI/SidePanel.cpp]: don't need to call `planet->Name()` again if you have a local copy / reference
alternatively, maybe you don't need the local copy / reference... just writing `planet->Name()` doesn't take up too much space

File: UI/SidePanel.cpp (hunk 6 above)
Comment (LGM-Doyle) [reply UI/SidePanel.cpp]: Fixed.

File: UI/BuildingsPanel.cpp (hunk 1 above)
Comment (LGM-Doyle) [reply UI/BuildingsPanel.cpp]: Fixed.

File: UI/BuildingsPanel.cpp (hunk 2 above)
Comment (LGM-Doyle) [reply UI/BuildingsPanel.cpp]: Fixed.

File: UI/SidePanel.cpp (hunk 3 above)
Comment (LGM-Doyle) [reply UI/SidePanel.cpp]: Fixed.

File: UI/SidePanel.cpp (hunk 4 above)
Comment (LGM-Doyle) [reply UI/SidePanel.cpp]: I used planet->Name() in both locations.

**Review by LGM-Doyle — COMMENTED** [review]
@adrianbroher, I agree some of them should not have been named.

I named all of the lambdas only because I used the same methodology for every `MenuItem` to move the code from the switch statement, into the menu item directly.  It seemed less error prone.

File: GG/src/Menu.cpp [hunk 7]
```diff
@@ -595,10 +310,11 @@ void PopupMenu::MouseHere(const Pt& pt, Flags<ModKey> mod_keys)
...
 {
     bool retval = Wnd::Run();
-    if (m_item_selected) {
-        (*m_item_selected->SelectedIDSignal)(m_item_selected->item_ID);
-        (*m_item_selected->SelectedSignal)();
-    }
+    if (retval
+        && m_item_selected
+        && m_item_selected->m_selected_on_close_callback)
+        m_item_selected->m_selected_on_close_callback();
```
Comment (geoffthemedio) [inline GG/src/Menu.cpp]: add { } after multi line if

File: UI/DesignWnd.cpp [hunk 8]
```diff
@@ -1835,50 +1835,44 @@ void BasesListBox::BaseRightClicked(GG::ListBox::iterator it, const GG::Pt& pt,
...
-        GG::MenuItem menu_contents;
-        if (client_empire_id != ALL_EMPIRES)
-            menu_contents.next_level.push_back(GG::MenuItem(UserString("DESIGN_OBSOLETE"), 1, false, false));
 
-        if (design->DesignedByEmpire() == client_empire_id)
-            menu_contents.next_level.push_back(GG::MenuItem(UserString("DESIGN_RENAME"), 2, false, false));
+        // Context menu actions
+        auto delete_design_action = [&client_empire_id, &design_id]() {
+            HumanClientApp::GetApp()->Orders().IssueOrder(
+                OrderPtr(new ShipDesignOrder(client_empire_id, design_id, true)));}
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: } to new line

File: UI/DesignWnd.cpp [hunk 9]
```diff
@@ -1895,35 +1889,25 @@ void BasesListBox::BaseRightClicked(GG::ListBox::iterator it, const GG::Pt& pt,
...
-            }
 
-            case 2: {   // add all saved designs
-                DebugLogger() << "BasesListBox::BaseRightClicked LoadAllSavedDesigns";
-                manager.LoadAllSavedDesigns();
-                break;
-            }
+        // Context menu actions
+        // add design
+        auto add_design_action = [&design, empire_id](){
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: space before {

File: UI/DesignWnd.cpp [hunk 10]
```diff
@@ -1895,35 +1889,25 @@ void BasesListBox::BaseRightClicked(GG::ListBox::iterator it, const GG::Pt& pt,
...
+            DebugLogger() << "BasesListBox::BaseRightClicked Add Saved Design" << design->Name();
+            int new_design_id = HumanClientApp::GetApp()->GetNewDesignID();
+            HumanClientApp::GetApp()->Orders().IssueOrder(
+                OrderPtr(new ShipDesignOrder(empire_id, new_design_id, *design)));};
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: } to new line

File: UI/DesignWnd.cpp [hunk 11]
```diff
@@ -1895,35 +1889,25 @@ void BasesListBox::BaseRightClicked(GG::ListBox::iterator it, const GG::Pt& pt,
...
 
-            default:
-                break;
-            }
-        }
+        // add all saved designs
+        auto add_all_saved_designs_action = [&manager](){
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: space before {

File: UI/DesignWnd.cpp [hunk 12]
```diff
@@ -1895,35 +1889,25 @@ void BasesListBox::BaseRightClicked(GG::ListBox::iterator it, const GG::Pt& pt,
...
+            DebugLogger() << "BasesListBox::BaseRightClicked LoadAllSavedDesigns";
+            manager.LoadAllSavedDesigns();};
```
Comment (geoffthemedio) [inline UI/DesignWnd.cpp]: } to new line

File: UI/FleetWnd.cpp [hunk 13]
```diff
@@ -2554,46 +2554,18 @@ void FleetDetailPanel::ShipRightClicked(GG::ListBox::iterator it, const GG::Pt&
...
+    CUIPopupMenu popup(pt.x, pt.y);
 
-    // create popup menu with a rename ship option in it.
-    GG::MenuItem menu_contents;
-
+    // Zoom to design
     if (design) {
         std::string popup_label = boost::io::str(FlexibleFormat(UserString("ENC_LOOKUP")) % design->Name(true));
-        menu_contents.next_level.push_back(GG::MenuItem(popup_label, 5, false, false));
+        auto zoom_to_design_action = [design]() {ClientUI::GetClientUI()->ZoomToShipDesign(design->ID()); };
```
Comment (geoffthemedio) [inline UI/FleetWnd.cpp]: space after {

File: UI/LinkText.cpp [hunk 14]
```diff
@@ -78,31 +78,19 @@ void LinkText::LClick(const GG::Pt& pt, GG::Flags<GG::ModKey> mod_keys)
 { TextLinker::LClick_(pt, mod_keys); }
 
 void LinkText::RClick(const GG::Pt& pt, GG::Flags<GG::ModKey> mod_keys) {
+
+    auto rclick_action = [this, pt, mod_keys]() { TextLinker::RClick_(pt, mod_keys); };
+    auto copy_action = [this] () { GG::GUI::GetGUI()->CopyWndText(this); };
```
Comment (geoffthemedio) [inline UI/LinkText.cpp]: no space before () I think... looks weird now

File: UI/ObjectListWnd.cpp [hunk 15]
```diff
@@ -2411,86 +2420,40 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
...
         }
-        GG::MenuItem focusMenuItem(UserString("MENUITEM_SET_FOCUS"), 3, false, false);
+        GG::MenuItem focusMenuItem(UserString("MENUITEM_SET_FOCUS"), false, false/*, no action*/);
         for (std::map<std::string, int>::value_type& entry : all_foci) {
             menuitem_id++;
+            auto focus_action = [this, entry, app, &focus_ship_building_common_action]() {
+                std::string focus = entry.first;
+                for (const auto& selection : m_list_box->Selections()) {
+                    ObjectRow *row = dynamic_cast<ObjectRow *>(*selection);
+                    if (row) {
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]: prefer if (!row) continue and de-indent, but not a big deal in this case

File: UI/ObjectListWnd.cpp [hunk 16]
```diff
@@ -2411,86 +2420,40 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
...
             }
         }
-        GG::MenuItem focusMenuItem(UserString("MENUITEM_SET_FOCUS"), 3, false, false);
+        GG::MenuItem focusMenuItem(UserString("MENUITEM_SET_FOCUS"), false, false/*, no action*/);
         for (std::map<std::string, int>::value_type& entry : all_foci) {
             menuitem_id++;
+            auto focus_action = [this, entry, app, &focus_ship_building_common_action]() {
+                std::string focus = entry.first;
+                for (const auto& selection : m_list_box->Selections()) {
+                    ObjectRow *row = dynamic_cast<ObjectRow *>(*selection);
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]: move * to type not variable name
remove space before * in <>

File: UI/ObjectListWnd.cpp [hunk 17]
```diff
@@ -2411,86 +2420,40 @@ void ObjectListWnd::ObjectRightClicked(GG::ListBox::iterator it, const GG::Pt& p
...
-                         }
-                    }
-                }
-            } else if (id > MENUITEM_SET_SHIP_BASE && id <= ship_menuitem_id) {
-                int ship_design = std::next(avail_designs.begin(), id - MENUITEM_SET_SHIP_BASE - 1)->first;
+            auto produce_ship_action = [this, it, app, cur_empire, &focus_ship_building_common_action]() {
+                int ship_design = it->first;
                 bool needs_queue_update(false);
                 for (const GG::ListBox::SelectionSet::value_type& entry : m_list_box->Selections()) {
                     ObjectRow *row = dynamic_cast<ObjectRow *>(*entry);
```
Comment (geoffthemedio) [inline UI/ObjectListWnd.cpp]: prefer * row to *row, no space before * in <>

File: UI/PlayerListWnd.cpp [hunk 18]
```diff
@@ -728,7 +728,23 @@ void PlayerListWnd::PlayerRightClicked(GG::ListBox::iterator it, const GG::Pt& p
         return;
     }
 
-    GG::MenuItem menu_contents;
+    auto make_send_diplomatic_action = [client_player_id, clicked_player_id, client_empire_id, clicked_empire_id](const std::function<DiplomaticMessage(int, int)>& message){
+        return [client_player_id, clicked_player_id, client_empire_id, clicked_empire_id, &message](){
```
Comment (geoffthemedio) [inline UI/PlayerListWnd.cpp]: space before {

File: UI/PlayerListWnd.cpp [hunk 19]
```diff
@@ -728,7 +728,23 @@ void PlayerListWnd::PlayerRightClicked(GG::ListBox::iterator it, const GG::Pt& p
...
+            HumanClientApp::GetApp()->Networking().SendMessage(
+                DiplomacyMessage(client_player_id, clicked_player_id, message(client_empire_id, clicked_empire_id)));};};
```
Comment (geoffthemedio) [inline UI/PlayerListWnd.cpp]: } to newline

File: UI/PlayerListWnd.cpp [hunk 20]
```diff
@@ -728,7 +728,23 @@ void PlayerListWnd::PlayerRightClicked(GG::ListBox::iterator it, const GG::Pt& p
...
+    // Actions
+    auto war_declaration_action = make_send_diplomatic_action(WarDeclarationDiplomaticMessage);
+    auto peace_proposal_action = make_send_diplomatic_action(PeaceProposalDiplomaticMessage);
+    auto peace_accept_action = make_send_diplomatic_action(AcceptPeaceDiplomaticMessage);
+    auto allies_proposal_action = make_send_diplomatic_action(AlliesProposalDiplomaticMessage);
+    auto allies_accept_action = make_send_diplomatic_action(AcceptAlliesDiplomaticMessage);
+    auto end_alliance_declaration_action = make_send_diplomatic_action(EndAllianceDiplomaticMessage);
+    auto proposal_cancel_action = make_send_diplomatic_action(CancelDiplomaticMessage);
+    auto proposal_reject_action = make_send_diplomatic_action(RejectProposalDiplomaticMessage);
+    auto pedia_lookup_action = [clicked_empire_id]() {ClientUI::GetClientUI()->ZoomToEmpire(clicked_empire_id);};
```
Comment (geoffthemedio) [inline UI/PlayerListWnd.cpp]: space after { and before }

File: UI/QueueListBox.cpp [hunk 21]
```diff
@@ -188,38 +188,34 @@ void QueueListBox::Clear() {
     DragDropLeave();
 }
 
+std::function<void()> QueueListBox::MoveToTopAction(GG::ListBox::iterator it) const
+{
```
Comment (geoffthemedio) [inline UI/QueueListBox.cpp]: { to previous line

File: UI/QueueListBox.cpp [hunk 22]
```diff
@@ -188,38 +188,34 @@ void QueueListBox::Clear() {
...
+    return [it, this]() {
+        if (GG::ListBox::Row* row = *it)
```
Comment (geoffthemedio) [inline UI/QueueListBox.cpp]: shouldn't this be an explicit dynamic_pointer_cast or similar?

File: UI/QueueListBox.cpp [hunk 23]
```diff
@@ -188,38 +188,34 @@ void QueueListBox::Clear() {
...
+            QueueItemMovedSignal(row, 0);
+    };
+}
+
+std::function<void()> QueueListBox::MoveToBottomAction(GG::ListBox::iterator it) const
+{
```
Comment (geoffthemedio) [inline UI/QueueListBox.cpp]: { to previous line

File: UI/SitRepPanel.cpp [hunk 24]
```diff
@@ -652,78 +664,43 @@ void SitRepPanel::DismissalMenu(GG::ListBox::iterator it, const GG::Pt& pt, cons
...
+            return;
         GG::GUI::GetGUI()->SetClipboardText(GG::Font::StripTags(sitrep_text));
-        break;
-    }
-    case 10: { // Display help article
-        ClientUI::GetClientUI()->ZoomToEncyclopediaEntry("SITREP_IGNORE_BLOCK_TITLE");
-        break;
-    }
+    };
+    auto help_action = []() {ClientUI::GetClientUI()->ZoomToEncyclopediaEntry("SITREP_IGNORE_BLOCK_TITLE");};
```
Comment (geoffthemedio) [inline UI/SitRepPanel.cpp]: space after { and before }

File: UI/TechTreeWnd.cpp [hunk 25]
```diff
@@ -1936,30 +1923,16 @@ void TechTreeWnd::TechListBox::TechRightClicked(GG::ListBox::iterator it, const
     if (rq.find(tech_name) != rq.end())
         return;
 
-    GG::MenuItem menu_contents;
+    auto tech_dclick_action = [this, it, pt]() {TechDoubleClicked(it, pt, GG::Flags<GG::ModKey>()); };
```
Comment (geoffthemedio) [inline UI/TechTreeWnd.cpp]: space after {

**Comment by geoffthemedio** [comment]
otherwise OK by me for merge

**Review by LGM-Doyle — COMMENTED** [review]
I made all the changes and I will merge.

**Pull request merged**